iree/compiler/InputConversion/MHLO/MHLOToLinalgOnTensors.cpp

```diff
@@ -268,6 +268,101 @@ struct ConstOpConversion : public OpConversionPattern<mhlo::ConstOp> {
   }
 };
 
+//===----------------------------------------------------------------------===//
+// mhlo.RngUniformOp conversion patterns.
+//===----------------------------------------------------------------------===//
+
+namespace {
+/// Returns an ArrayAttr that contains `nLoops` attributes. All the attributes
+/// are "parallel" except the last `nReduction` elements, where are "reduction"
+/// attributes.
+SmallVector<StringRef, 3> GetParallelAndReductionIterators(
+    unsigned nLoops, unsigned nReduction) {
+  SmallVector<StringRef, 3> res(nLoops - nReduction,
+                                getParallelIteratorTypeName());
+  res.append(nReduction, getReductionIteratorTypeName());
+  return res;
+}
+
+struct RngUniformConversion : public OpConversionPattern<mhlo::RngUniformOp> {
+  using OpConversionPattern<mhlo::RngUniformOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::RngUniformOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const final {
+    Type int32Type = IntegerType::get(op.getContext(), /*width=*/32);
+    auto loc = op.getLoc();
+    auto targetTy = op->getResult(0).getType().dyn_cast<ShapedType>();
+    auto targetRank = targetTy.getRank();
+    Value output;
+    Value initTensor = rewriter.create<linalg::InitTensorOp>(
```

Comment:
This is assuming that this is statically shaped. Can we just support dynamic shapes here too. All you need to do is insert `dim` operations for the dimensions that are not known statically.

Reply:
hmm, I am trying to add dynamic size support but can't seem to make a proper example, since initTensorOp is technically using the contents/element value within the input tensor "shape". which should technically be real values.

Reply:
Ah I see the reason. The shape information is a tensor, ie a list of tensors. E.g.,

```
func @rng_uniform_3d(%min: tensor<f32>, %max: tensor<f32>) -> tensor<2x2x2xf32>
{
        %shape = constant dense<[2, 2, 2]>  : tensor<3xi32>
        %0 = "mhlo.rng_uniform"(%min, %max, %shape) : (tensor<f32>, tensor<f32>, tensor<3xi32>) -> tensor<2x2x2xf32>
        return %0 : tensor<2x2x2xf32>
}
```

The shape is `2x2x2` where is derived from `constant dense<[2, 2, 2]>  : tensor<3xi32>`.  I think what we want here is `tensor.extract` the shape tensor. Not `tensor.dim` operations.

Reply:
yes :), currently I am taking the shape from the output of mhlo.rng_uniform itself. but even if I use tensor extractOp, does there exist dynamic value as element of tensor haha S.T we can check with ShapedType::kDynamicSize?

Reply:
Yes, we can check with `ShapedType::kDynamicSize`. I think we can add the test like:

```
func @rng_uniform_3d(%min: tensor<f32>, %max: tensor<f32>, %shape: tensor<1xf32>) -> tensor<?xf32>
{
        %0 = "mhlo.rng_uniform"(%min, %max, %shape) : (tensor<f32>, tensor<f32>, tensor<3xi32>) -> tensor<?xf32>
        return %0 : tensor<?xf32>
}
```

Reply:
I see I see, so in this case %shape is never really `constant dense<[?]>`, but more along the line that we are not sure a head of time what value %shape will actually be yet?

Reply:
Yes, I think so. Based on the operation semantics, mhlo.rng_uniform takes a shape operand. In tensors world, we can extract the value by inserting `tensor.extract`.

To test the dynamic case, it will never be `constant dense<[?]>`. Because constant is a known value. So I'd test with a function input. I'm not saying that this is the case we will see for dynamic cases. But this is one of ways to test it and it looks reasonable to me (which still satisfies the operation semantics).

Reply:
Tested with iree-run-mlir, seems to dynamic size seems work well now, thanks Hanhan! :)